src/filecache.c

```diff
@@ -1504,19 +1578,38 @@ void filecache_forensic_haven(const char *cache_path, filecache_t *cache, const
     asprintf(&buf, "path: %s\ncache filename: %s\nlast_server_update: %lu\nfilesize: %lu\nfailed_rename %d\n",
         path, pdata->filename, pdata->last_server_update, fsize, failed_rename);
     bytes_written = write(fd, buf, strlen(buf));
-    log_print(LOG_DEBUG, SECTION_FILECACHE_CACHE, "filecache_forensic_haven: write (%s) of fd %d returns %d", newpath, fd, bytes_written);
+    log_print(LOG_DEBUG, SECTION_FILECACHE_CACHE, "%s: write (%s) of fd %d returns %d", fname, newpath, fd, bytes_written);
     if (bytes_written < 0) {
-        log_print(LOG_NOTICE, SECTION_FILECACHE_CACHE, "filecache_forensic_haven: error on write (%s) of fd %d returns %d", newpath, fd, bytes_written);
-        g_set_error(gerr, filecache_quark(), errno, "filecache_forensic_haven: Failed on write to %s", newpath);
+        log_print(LOG_NOTICE, SECTION_FILECACHE_CACHE, "%s: error on write (%s) of fd %d returns %d", fname, newpath, fd, bytes_written);
+        g_set_error(gerr, filecache_quark(), errno, "%s: Failed on write to %s", fname, newpath);
         goto finish;
     }
 
 finish:
     if (fd >= 0) close(fd);
     free(buf);
     free(newpath);
+    newpath = NULL;
     free(pdata);
-    log_print(LOG_DEBUG, SECTION_FILECACHE_CACHE, "filecache_forensic_haven: exiting for %s", path);
+    // Cleanup older entries in the forensic haven
+    asprintf(&newpath, "%s/%s/", cache_path, forensic_haven_dir);
+    // Clear out all files older than a day
+    while (files_left >= files_kept && hours > 0) {
+        files_left = clear_files(newpath, time(NULL) - (hours * 60 * 60), &subgerr);
+        if (subgerr) {
+            log_print(LOG_ERR, SECTION_FILECACHE_FILE, 
+                    "%s: error on clear_files: %s -- %d: %s", 
+                    fname, subgerr->message, subgerr->code, g_strerror(subgerr->code));
+            break;
+        }
+        hours /= 4;
+        log_print(LOG_NOTICE, SECTION_FILECACHE_FILE, 
+                "%s: files_left: %d; files_kept: %d; hours: %d", 
+                fname, files_left, files_kept, hours);
+    }
+    free(newpath);
```

Comment:
We free a previous alloc. Then we realloc via asprintf, then we free again. At least that's the theory.